src/NUnitFramework/framework/Internal/Randomizer.cs

```diff
@@ -492,34 +495,127 @@ public T NextEnum<T>()
 
         /// <summary>
-        /// Generate a random string based on the characters from the input string.
+        /// Returns a random string of the specified length using characters from the specified set.
         /// </summary>
-        /// <param name="outputLength">desired length of output string.</param>
-        /// <param name="allowedChars">string representing the set of characters from which to construct the resulting string</param>
-        /// <returns>A random string of arbitrary length</returns>
+        /// <param name="outputLength">The desired length of the string in char units</param>
+        /// <param name="allowedChars">The characters to use in building the string</param>
+        /// <returns>A random string</returns>
+        /// <exception cref="ArgumentException">
+        /// Thrown if allowedChars contains surrogate characters that would create invalid UTF-16 strings,
+        /// or if the requested outputLength cannot be satisfied with the provided characters
+        /// </exception>
         public string GetString(int outputLength, string allowedChars)
         {
-            if (outputLength < 0)
-                throw new ArgumentOutOfRangeException(nameof(outputLength));
+            // Fast path: Check if allowedChars contains any surrogates
+            bool hasSurrogates = false;
+            foreach (var c in allowedChars)
+            {
+                if (char.IsSurrogate(c))
+                {
+                    hasSurrogates = true;
+                    break;
+                }
+            }
+
+            if (hasSurrogates)
+            {
+                return GetStringSafe(outputLength, allowedChars);
+            }
+
 #if NET6_0_OR_GREATER
-            return string.Create(outputLength, allowedChars, FillSpan);
+            return GetStringUnsafeSpan(outputLength, allowedChars);
+#else
+            return GetStringUnsafeArray(outputLength, allowedChars);
+#endif
+        }
 
-            void FillSpan(Span<char> data, string allowedChars)
+#if NET6_0_OR_GREATER
+        private string GetStringUnsafeSpan(int outputLength, string allowedChars)
+        {
+            // Use stackalloc for small strings, heap for large ones
+            const int maxStackLimit = 256; // Reasonable stack limit
+
+            if (outputLength <= maxStackLimit)
             {
+                Span<char> data = stackalloc char[outputLength];
                 for (int i = 0; i < data.Length; i++)
                     data[i] = allowedChars[Next(0, allowedChars.Length)];
+                return new string(data);
             }
-#else
-            var data = new char[outputLength];
 
+            // Fall back to array for very large strings
+            return GetStringUnsafeArray(outputLength, allowedChars);
+        }
+#endif
+
+        private string GetStringUnsafeArray(int outputLength, string allowedChars)
+        {
+            var data = new char[outputLength];
```

Comment:
request: If we prefer to avoid the closure of sting.Create(), could we instead then use ArrayPool on non-netfx ?

Something like (untested):
```csharp
            char[]? data = null;
            try
            {
#if !NETFRAMEWORK
                data = ArrayPool<char>.Shared.Rent(outputLength);
#else
                data = new char[outputLength];
#endif
                for (int i = 0; i < outputLength; i++)
                    data[i] = allowedChars[Next(0, allowedChars.Length)];
                return new string(data, 0, outputLength);
            }
            finally
            {
#if !NETFRAMEWORK
                if (data is not null)
                {
                    ArrayPool<char>.Shared.Return(data);
                }
#endif
            }
```

Reply:
Probably, but I am on thin ice here.  If you would like to change this in this PR, I am just happy to see that.  
I assume it will give a performance boost, right ?

Reply:
It "should" give a boost. I'd be happy to push a commit with that change in a bit after I benchmark it to confirm